### PR TITLE
feat(vision): add Vision DP for parallel ViT computation across Ulysses SP ranks

### PR DESCRIPTION
**Pull request by aoshen524**
## Vision Data Parallel: Distribute ViT computation across Ulysses SP ranks

Ported from [verl PR #5230](https://github.com/verl-project/verl/pull/5230), adapted for ROLL's Ulysses SP infrastructure.

### Motivation

When using Ulysses Sequence Parallelism (sp_size > 1), the VisionTransformer still processes ALL images on every rank, wasting memory. Vision DP distributes whole images across SP ranks, reducing ViT memory by ~sp_size x.

### Key changes

| File | Change |
|------|--------|
| `roll/utils/context_parallel/vision_dp.py` | Core utilities: load-balanced assignment, tensor slicing, all-gather with gradient fix |
| `roll/utils/context_parallel/monkey_patch.py` | Integration with idempotency guard, clean unapply support |
| `tests/utils/test_vision_dp_on_cpu.py` | CPU-only unit tests (28 tests) |

### Tests

```bash
python -m pytest tests/utils/test_vision_dp_on_cpu.py -v
# 28 passed
```

### CODE REVIEW

File: roll/distributed/strategy/deepspeed_strategy.py [hunk 1]
```diff
@@ -332,6 +334,8 @@ def initialize(self, model_provider):
         if (cp_size := self.worker_config.model_args.ulysses_size) > 1:
             current_platform.apply_ulysses_patch()
             set_upg_manager(ulysses_size=cp_size, rank=global_rank, world_size=world_size)
+            if self.worker_config.model_args.vision_dp:
+                apply_vision_dp_patch()
```
Comment (guoshengCS) [inline roll/distributed/strategy/deepspeed_strategy.py]: It seems vision_dp also suit to fsdp strategy in the same way, can you please support it in fsdp_strategy

**Comment by guoshengCS** [comment]
@aoshen524 Could you please provide training curves with and without vision DP

File: roll/utils/context_parallel/vision_dp.py [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+            else:
+                # Empty rank: create matching empty deepstack tensors
+                num_deepstack = len(self.deepstack_merger_list)
+                h = local_embeddings.shape[1]
+                local_deepstack = [
+                    torch.empty(
+                        (0, h), dtype=hidden_states.dtype, device=hidden_states.device
+                    )
+                    for _ in range(num_deepstack)
+                ]
```
Comment (pUmpKin-Co) [inline roll/utils/context_parallel/vision_dp.py]: Why doesn't the empty-rank local_deepstack path also call `requires_grad_()`, similar to empty local_embeddings? Since each deepstack tensor is also passed through `GatherVisionEmbeddings`, it seems those empty tensors should still participate in autograd so every rank enters the same backward all_reduce. Otherwise, could empty ranks skip the custom backward for deepstack and risk a collective mismatch or hang?
